lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm

```diff
@@ -386,13 +386,15 @@ sub connect_to_cmd_srv {
             # upgrade to ws connection if not already a websocket connection
             if (!$tx->is_websocket) {
                 my $location_header = ($tx->completed ? $tx->res->headers->location : undef);
-                if (!$location_header) {
+                if (!$location_header && !$tx->res->headers->upgrade) {
```

Comment:
Not sure how costly `$tx->res->headers` is but I'd save the headers in a variable instead of calling it twice.

Reply:
I'm also not sure whether this makes actually sense in all cases. Maybe we need to follow a normal redirection before we end up being able to upgrade? Of course it would make sense to limit the number of redirections to follow. Or does this Mojo for us.

---

@kraih Any suggestion or best practices for this? I actually just copied that code from our worker.

---

Note that the code I've copied from is (at this point) in `OpenQA::Worker::WebUIConnection::_setup_websocket_connection` and likely should be adjusted as well.

Reply:
I generate wrappers, but those just start the services with custom OPENQA_BASE_PORT defined and all rest is default (i.e. no config file), i.e.:
`OPENQA_BASE_PORT=${port} src/script/openqa daemon`
`OPENQA_BASE_PORT=${port} src/script/openqa-livehandler daemon`
`src/script/worker --isotovideo 'os-autoinst/isotovideo' --host http://127.0.0.1:${port} --instance 1 --apikey 1234567890ABCDEF --apisecret 1234567890ABCDEF`
etc . Then use `job post`. 
Basically you can just execute this script and look into few-lines start.sh files to see that https://github.com/andrii-suse/environs/blob/master/.product/qa/branch/.example/fullstack.sh

Reply:
I don't understand this code at all. The whole `if (!$tx->is_websocket) { ` block makes no sense to me.

Reply:
When I do not set `WORKER_HOSTNAME =` in `wokers.ini`,  the error message `unable to upgrade ws to command server` will be shown in webUI. I am not sure if this situation is the same with yours.

Reply:
fwiw. Redirecting WebSockets is an optional feature in RFC 6455, and Mojolicious does not support it.

Reply:
@kraih So Mojolicious should handle this automatically and follow redirections as needed? I didn't write this block initially (just copied it from the worker) but I *suspect* the motivation behind this block is to catch the case when there's just a regular HTTP server reachable and no upgrade is possible (e.g. the web socket server is not running so one just gets the error message from the reverse proxy). I also remember having errors when calling `send()` because when the upgrade is not possible one just ends up with a regular tx and that case needs to be distinguished. The `if (!$tx->is_websocket) {` block prevents that.

Reply:
@Amrysliu thanks you, that may be the case - I did provide `WORKER_HOSTNAME=127.0.0.1` and `WORKER_HOSTNAME=localhost` as parameter to `job post`, I shall try ini file as well.

Reply:
@andrii-suse This settings *needs* to go into the worker config file. It is a worker configuration and not a job setting.

Reply:
But as it turns out 307/308 redirects work with Mojolicious by accident, so i [turned that into a feature](https://github.com/mojolicious/mojo/compare/8fc68f476968...375e86134e4c) upstream. And that means you can just set `$self->app->ua->max_redirects(10)` and let `Mojo::UserAgent` worry about redirects and limit enforcement.

Reply:
Updated the comment above to clarify that all redirects are now supported for WebSockets with `Mojo::UserAgent`.